lib/cachebs/cachebs.go

```diff
@@ -0,0 +1,80 @@
+package cachebs
+
+import (
+	"context"
+
+	lru "github.com/hashicorp/golang-lru"
+	block "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	bstore "github.com/ipfs/go-ipfs-blockstore"
+	logging "github.com/ipfs/go-log/v2"
+)
+
+var log = logging.Logger("cachebs")
+
+type CacheBS struct {
+	cache *lru.ARCCache
+	bs    bstore.Blockstore
+}
+
+func NewBufferedBstore(base blockstore.Blockstore, size int) *CacheBS {
+	c, err := lru.NewARC(size)
+	if err != nil {
+		panic(err)
+	}
+	return &CacheBS{
+		cache: c,
+		bs:    base,
+	}
+}
+
+var _ (bstore.Blockstore) = &CacheBS{}
+
+func (bs *CacheBS) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
+	return bs.bs.AllKeysChan(ctx)
+}
+
+func (bs *CacheBS) DeleteBlock(c cid.Cid) error {
+	bs.cache.Remove(c)
+
+	return bs.bs.DeleteBlock(c)
+}
+
+func (bs *CacheBS) Get(c cid.Cid) (block.Block, error) {
+	v, ok := bs.cache.Get(c)
+	if ok {
+		return v.(block.Block), nil
+	}
+
+	return bs.bs.Get(c)
```

Comment:
After calling `Get` to underlying blockstore we should call `Add` to local cache, otherwise the cache will not populate with older items.